// bybit-api-gnome/lib/util/websocket-util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.neverGuard = exports.WS_ERROR_ENUM = exports.getUsdcWsKeyForTopic = exports.getMaxTopicsPerSubscribeEvent = exports.getWsKeyForTopic = exports.PUBLIC_WS_KEYS = exports.WS_AUTH_ON_CONNECT_KEYS = exports.WS_KEY_MAP = exports.WS_BASE_URL_MAP = void 0;
exports.WS_BASE_URL_MAP = {
    inverse: {
        public: {
            livenet: 'wss://stream.bybit.com/realtime',
            testnet: 'wss://stream-testnet.bybit.com/realtime',
        },
        private: {
            livenet: 'wss://stream.bybit.com/realtime',
            testnet: 'wss://stream-testnet.bybit.com/realtime',
        },
    },
    linear: {
        public: {
            livenet: 'wss://stream.bybit.com/realtime_public',
            livenet2: 'wss://stream.bytick.com/realtime_public',
            testnet: 'wss://stream-testnet.bybit.com/realtime_public',
        },
        private: {
            livenet: 'wss://stream.bybit.com/realtime_private',
            livenet2: 'wss://stream.bytick.com/realtime_private',
            testnet: 'wss://stream-testnet.bybit.com/realtime_private',
        },
    },
    spot: {
        public: {
            livenet: 'wss://stream.bybit.com/spot/quote/ws/v1',
            livenet2: 'wss://stream.bybit.com/spot/quote/ws/v2',
            testnet: 'wss://stream-testnet.bybit.com/spot/quote/ws/v1',
            testnet2: 'wss://stream-testnet.bybit.com/spot/quote/ws/v2',
        },
        private: {
            livenet: 'wss://stream.bybit.com/spot/ws',
            testnet: 'wss://stream-testnet.bybit.com/spot/ws',
        },
    },
    spotv3: {
        public: {
            livenet: 'wss://stream.bybit.com/spot/public/v3',
            testnet: 'wss://stream-testnet.bybit.com/spot/public/v3',
        },
        private: {
            livenet: 'wss://stream.bybit.com/spot/private/v3',
            testnet: 'wss://stream-testnet.bybit.com/spot/private/v3',
        },
    },
    usdcOption: {
        public: {
            livenet: 'wss://stream.bybit.com/trade/option/usdc/public/v1',
            livenet2: 'wss://stream.bytick.com/trade/option/usdc/public/v1',
            testnet: 'wss://stream-testnet.bybit.com/trade/option/usdc/public/v1',
        },
        private: {
            livenet: 'wss://stream.bybit.com/trade/option/usdc/private/v1',
            livenet2: 'wss://stream.bytick.com/trade/option/usdc/private/v1',
            testnet: 'wss://stream-testnet.bybit.com/trade/option/usdc/private/v1',
        },
    },
    usdcPerp: {
        public: {
            livenet: 'wss://stream.bybit.com/perpetual/ws/v1/realtime_public',
            livenet2: 'wss://stream.bytick.com/perpetual/ws/v1/realtime_public',
            testnet: 'wss://stream-testnet.bybit.com/perpetual/ws/v1/realtime_public',
        },
        private: {
            livenet: 'wss://stream.bybit.com/trade/option/usdc/private/v1',
            livenet2: 'wss://stream.bytick.com/trade/option/usdc/private/v1',
            testnet: 'wss://stream-testnet.bybit.com/trade/option/usdc/private/v1',
        },
    },
    unifiedOption: {
        public: {
            livenet: 'wss://stream.bybit.com/option/usdc/public/v3',
            testnet: 'wss://stream-testnet.bybit.com/option/usdc/public/v3',
        },
        private: {
            livenet: 'wss://stream.bybit.com/unified/private/v3',
            testnet: 'wss://stream-testnet.bybit.com/unified/private/v3',
        },
    },
    unifiedPerp: {
        public: {
            livenet: 'useBaseSpecificEndpoint',
            testnet: 'useBaseSpecificEndpoint',
        },
        private: {
            livenet: 'wss://stream.bybit.com/unified/private/v3',
            testnet: 'wss://stream-testnet.bybit.com/unified/private/v3',
        },
    },
    unifiedPerpUSDT: {
        public: {
            livenet: 'wss://stream.bybit.com/contract/usdt/public/v3',
            testnet: 'wss://stream-testnet.bybit.com/contract/usdt/public/v3',
        },
        private: {
            livenet: 'useUnifiedEndpoint',
            testnet: 'useUnifiedEndpoint',
        },
    },
    unifiedPerpUSDC: {
        public: {
            livenet: 'wss://stream.bybit.com/contract/usdc/public/v3',
            testnet: 'wss://stream-testnet.bybit.com/contract/usdc/public/v3',
        },
        private: {
            livenet: 'useUnifiedEndpoint',
            testnet: 'useUnifiedEndpoint',
        },
    },
    contractUSDT: {
        public: {
            livenet: 'wss://stream.bybit.com/contract/usdt/public/v3',
            testnet: 'wss://stream-testnet.bybit.com/contract/usdt/public/v3',
        },
        private: {
            livenet: 'wss://stream.bybit.com/contract/private/v3',
            testnet: 'wss://stream-testnet.bybit.com/contract/private/v3',
        },
    },
    contractInverse: {
        public: {
            livenet: 'wss://stream.bybit.com/contract/inverse/public/v3',
            testnet: 'wss://stream-testnet.bybit.com/contract/inverse/public/v3',
        },
        private: {
            livenet: 'wss://stream.bybit.com/contract/private/v3',
            testnet: 'wss://stream-testnet.bybit.com/contract/private/v3',
        },
    },
};
exports.WS_KEY_MAP = {
    inverse: 'inverse',
    linearPrivate: 'linearPrivate',
    linearPublic: 'linearPublic',
    spotPrivate: 'spotPrivate',
    spotPublic: 'spotPublic',
    spotV3Private: 'spotV3Private',
    spotV3Public: 'spotV3Public',
    usdcOptionPrivate: 'usdcOptionPrivate',
    usdcOptionPublic: 'usdcOptionPublic',
    usdcPerpPrivate: 'usdcPerpPrivate',
    usdcPerpPublic: 'usdcPerpPublic',
    unifiedPrivate: 'unifiedPrivate',
    unifiedOptionPublic: 'unifiedOptionPublic',
    unifiedPerpUSDTPublic: 'unifiedPerpUSDTPublic',
    unifiedPerpUSDCPublic: 'unifiedPerpUSDCPublic',
    contractUSDTPublic: 'contractUSDTPublic',
    contractUSDTPrivate: 'contractUSDTPrivate',
    contractInversePublic: 'contractInversePublic',
    contractInversePrivate: 'contractInversePrivate',
};
exports.WS_AUTH_ON_CONNECT_KEYS = [
    exports.WS_KEY_MAP.spotV3Private,
    exports.WS_KEY_MAP.usdcOptionPrivate,
    exports.WS_KEY_MAP.usdcPerpPrivate,
    exports.WS_KEY_MAP.unifiedPrivate,
    exports.WS_KEY_MAP.contractUSDTPrivate,
    exports.WS_KEY_MAP.contractInversePrivate,
];
exports.PUBLIC_WS_KEYS = [
    exports.WS_KEY_MAP.linearPublic,
    exports.WS_KEY_MAP.spotPublic,
    exports.WS_KEY_MAP.spotV3Public,
    exports.WS_KEY_MAP.usdcOptionPublic,
    exports.WS_KEY_MAP.usdcPerpPublic,
    exports.WS_KEY_MAP.unifiedOptionPublic,
    exports.WS_KEY_MAP.unifiedPerpUSDTPublic,
    exports.WS_KEY_MAP.unifiedPerpUSDCPublic,
    exports.WS_KEY_MAP.contractUSDTPublic,
    exports.WS_KEY_MAP.contractInversePublic,
];
/** Used to automatically determine if a sub request should be to the public or private ws (when there's two) */
const PRIVATE_TOPICS = [
    'position',
    'execution',
    'order',
    'stop_order',
    'wallet',
    'outboundAccountInfo',
    'executionReport',
    'ticketInfo',
    // copy trading apis
    'copyTradePosition',
    'copyTradeOrder',
    'copyTradeExecution',
    'copyTradeWallet',
    // usdc options
    'user.openapi.option.position',
    'user.openapi.option.trade',
    'user.order',
    'user.openapi.option.order',
    'user.service',
    'user.openapi.greeks',
    'user.mmp.event',
    // usdc perps
    'user.openapi.perp.position',
    'user.openapi.perp.trade',
    'user.openapi.perp.order',
    'user.service',
    // unified margin
    'user.position.unifiedAccount',
    'user.execution.unifiedAccount',
    'user.order.unifiedAccount',
    'user.wallet.unifiedAccount',
    'user.greeks.unifiedAccount',
    // contract v3
    'user.position.contractAccount',
    'user.execution.contractAccount',
    'user.order.contractAccount',
    'user.wallet.contractAccount',
];
function getWsKeyForTopic(market, topic, isPrivate) {
    const isPrivateTopic = isPrivate === true || PRIVATE_TOPICS.includes(topic);
    switch (market) {
        case 'inverse': {
            return exports.WS_KEY_MAP.inverse;
        }
        case 'linear': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.linearPrivate
                : exports.WS_KEY_MAP.linearPublic;
        }
        case 'spot': {
            return isPrivateTopic ? exports.WS_KEY_MAP.spotPrivate : exports.WS_KEY_MAP.spotPublic;
        }
        case 'spotv3': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.spotV3Private
                : exports.WS_KEY_MAP.spotV3Public;
        }
        case 'usdcOption': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.usdcOptionPrivate
                : exports.WS_KEY_MAP.usdcOptionPublic;
        }
        case 'usdcPerp': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.usdcPerpPrivate
                : exports.WS_KEY_MAP.usdcPerpPublic;
        }
        case 'unifiedOption': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.unifiedPrivate
                : exports.WS_KEY_MAP.unifiedOptionPublic;
        }
        case 'unifiedPerp': {
            if (isPrivateTopic) {
                return exports.WS_KEY_MAP.unifiedPrivate;
            }
            const upperTopic = topic.toUpperCase();
            if (upperTopic.indexOf('USDT') !== -1) {
                return exports.WS_KEY_MAP.unifiedPerpUSDTPublic;
            }
            if (upperTopic.indexOf('USDC') !== -1) {
                return exports.WS_KEY_MAP.unifiedPerpUSDCPublic;
            }
            throw new Error(`Failed to determine wskey for unified perps topic: "${topic}`);
        }
        case 'contractInverse': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.contractInversePrivate
                : exports.WS_KEY_MAP.contractInversePublic;
        }
        case 'contractUSDT': {
            return isPrivateTopic
                ? exports.WS_KEY_MAP.contractUSDTPrivate
                : exports.WS_KEY_MAP.contractUSDTPublic;
        }
        default: {
            throw neverGuard(market, `getWsKeyForTopic(): Unhandled market`);
        }
    }
}
exports.getWsKeyForTopic = getWsKeyForTopic;
function getMaxTopicsPerSubscribeEvent(market) {
    switch (market) {
        case 'inverse':
        case 'linear':
        case 'usdcOption':
        case 'usdcPerp':
        case 'unifiedOption':
        case 'unifiedPerp':
        case 'spot':
        case 'contractInverse':
        case 'contractUSDT': {
            return null;
        }
        case 'spotv3': {
            return 10;
        }
        default: {
            throw neverGuard(market, `getWsKeyForTopic(): Unhandled market`);
        }
    }
}
exports.getMaxTopicsPerSubscribeEvent = getMaxTopicsPerSubscribeEvent;
function getUsdcWsKeyForTopic(topic, subGroup) {
    const isPrivateTopic = PRIVATE_TOPICS.includes(topic);
    if (subGroup === 'option') {
        return isPrivateTopic
            ? exports.WS_KEY_MAP.usdcOptionPrivate
            : exports.WS_KEY_MAP.usdcOptionPublic;
    }
    return isPrivateTopic
        ? exports.WS_KEY_MAP.usdcOptionPrivate
        : exports.WS_KEY_MAP.usdcOptionPublic;
    // return isPrivateTopic
    //   ? WS_KEY_MAP.usdcPerpPrivate
    //   : WS_KEY_MAP.usdcPerpPublic;
}
exports.getUsdcWsKeyForTopic = getUsdcWsKeyForTopic;
exports.WS_ERROR_ENUM = {
    NOT_AUTHENTICATED_SPOT_V3: '-1004',
    API_ERROR_GENERIC: '10001',
    API_SIGN_AUTH_FAILED: '10003',
    USDC_OPTION_AUTH_FAILED: '3303006',
};
function neverGuard(x, msg) {
    return new Error(`Unhandled value exception "x", ${msg}`);
}
exports.neverGuard = neverGuard;
